### PR TITLE
Mldbfb 601 http client integration

### PR DESCRIPTION
**Pull request by wsourdeau**
This PR provides an implementation of HTTPClient from upstream, as a milestone towards MLDBFB-239. The curl-based implementation was chosen as its performance was sufficiently good for S3 and as it is considered to be the most stable one.
Additionally:
* some related classes have been fixed after bugs were revealed
* libc_value_description has been added

### CODE REVIEW

File: http/http_client.h [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+   - since those require header interpretation, there is not support for
+     cookies per se
+*/
+
+#pragma once
+
+#include <memory>
+#include <string>
+
+#include "mldb/ext/jsoncpp/value.h"
```
Comment (jeremybarnes) [inline http/http_client.h]: Would be good to not include the async_event_source header here, and to put more of the methods in the .cc file.

File: http/http_client.h [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+    RestParams headers_;
+    int timeout_;
+};
+
+
+/****************************************************************************/
+/* HTTP CLIENT IMPL                                                         */
+/****************************************************************************/
+
+struct HttpClientImpl : public AsyncEventSource {
```
Comment (jeremybarnes) [inline http/http_client.h]: Could we use composition rather than inheritance here?

File: http/http_client.h [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+};
+
+std::ostream & operator << (std::ostream & stream, HttpClientError error);
+
+
+/****************************************************************************/
+/* HTTP CLIENT                                                              */
+/****************************************************************************/
+
+struct HttpClient : public AsyncEventSource {
```
Comment (jeremybarnes) [inline http/http_client.h]: Same here.

File: http/http_client.h [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+
+    /* "baseUrl": scheme, hostname and port (scheme://hostname[:port]) that
+       will be used as base for all requests
+       "numParallels": number of requests that can be handled simultaneously
+       "queueSize": size of the backlog of pending requests, after which
+       operations will be refused (0 = infinite)
+       "implVersion": use version X of the HttpClientImpl, fallback
+       to HTTP_CLIENT_IMPL
+    */
+    HttpClient(const std::string & baseUrl,
```
Comment (jeremybarnes) [inline http/http_client.h]: Should all be in the .cc file.

File: http/http_client.h [hunk 5]
```diff
@@ -0,0 +1,432 @@
...
+
+
+/****************************************************************************/
+/* HTTP CLIENT SIMPLE CALLBACKS                                             */
+/****************************************************************************/
+
+/* This class is a child of HttpClientCallbacks and offers a simplified
+ * interface when support for progressive responses is not necessary. */
+
+struct HttpClientSimpleCallbacks : public HttpClientCallbacks
```
Comment (jeremybarnes) [inline http/http_client.h]: Why do we need a second class here?

File: http/tcp_socket_handler_impl.h [hunk 6]
```diff
@@ -64,6 +64,7 @@ struct TcpSocketHandlerImpl {
     typedef std::function<void(const boost::system::error_code & ec,
                                size_t bufferSize)> OnReadSome;
     OnReadSome onReadSome_;
+    bool closed_;
```
Comment (jeremybarnes) [inline http/tcp_socket_handler_impl.h]: atomic?

**Comment by jeremybarnes** [comment]
General comments
- Attempt to have as few includes and inlined methods as possible.  MLDB's compilation is slowed down by them and they tend to inroduce ABI couplings
- Please use Doxygen style /** comments for the method documentation, as Doxygen is used by the team to help generate code
- Please add a PR description (copied from the ticket is fine) as well as an indication as to what this will improve, and what the path forward is for further changesets that will come on top of this one

There were a few other minor comments in the code, but generally it seems to be in decent shape.

File: http/http_client.h (hunk 3 above)
Comment (wsourdeau) [inline http/http_client.h]: HttpClient is meant to be added to a MessageLoop (pending the asio version). In this case, it's required by the interface of MessageLoop::addSource.

File: http/http_client.h (hunk 5 above)
Comment (wsourdeau) [inline http/http_client.h]: HttpClientCallbacks is the full version that enables on-the-fly handling of requests and can therefore provide a version that is more memory efficient. The HttpClientSimpleCallbacks version enables easy handling of the requests through the need of a single callback with, in counterpart, the requirement to keep the full response body in memory.

File: http/http_client.h (hunk 3 above)
Comment (jeremybarnes) [inline http/http_client.h]: You could still decouple them, by having a getSource() method that returned a reference to an internal class that implemented MessageLoop::addSource.  The users of HttpClient shouldn't need to drag in files based upon the implementation.

File: http/http_client.h (hunk 5 above)
Comment (jeremybarnes) [inline http/http_client.h]: Why not make the HtpClientCallbacks implement the simple version by default?  Then we only have one class, and can override the other.

File: http/http_client.h (hunk 3 above)
Comment (wsourdeau) [inline http/http_client.h]: Good idea. Then I could require the constructor to take a MessageLoop as parameter, similarly to how the EventLoop class is used. Would take make sense?

File: http/http_client.h (hunk 5 above)
Comment (wsourdeau) [inline http/http_client.h]: I could be it does mix both point of views in the same class, is that desirable?

File: http/tcp_socket_handler_impl.h (hunk 6 above)
Comment (wsourdeau) [inline http/tcp_socket_handler_impl.h]: Indeed!

File: http/http_client.h (hunk 4 above)
Comment (wsourdeau) [inline http/http_client.h]: Done. I have segregated most classes in their own files and put all the code in the .cc. Should make things much clearer and reduce the number of required includes.

File: http/http_client.h (hunk 3 above)
Comment (jeremybarnes) [inline http/http_client.h]: Exactly.  And that parameter can be an opaque type.

**Comment by wsourdeau** [comment]
@jeremybarnes I have fixed the documentation for HttpClient, HttpClientImpl and LegacyEventLoop. Also, I have added a new module named "libc_value_description" with descriptions for struct provided by the libc (rusage, timeval).

If that looks fine to you, can I have your +1 ?

File: http/http_client_impl_v1.cc [hunk 7]
```diff
@@ -0,0 +1,669 @@
...
+    string headerLine(data, size);
+    if (headerLine.find("HTTP/1.1 100") == 0) {
+        afterContinue_ = true;
+    }
+    else if (afterContinue_) {
+        if (headerLine == "\r\n")
+            afterContinue_ = false;
+    }
+    else {
+        const auto & cbs = request_->callbacks();
```
Comment (FinchPowers) [inline http/http_client_impl_v1.cc]: Je suppose que cette fonction aussi est noexcept?

File: http/http_client_impl_v1.cc [hunk 8]
```diff
@@ -0,0 +1,669 @@
...
+        if (headerLine.find("HTTP/") == 0) {
+            size_t lineSize = headerLine.size();
+            size_t oldTokenIdx(0);
+            size_t tokenIdx = headerLine.find(" ");
+            if (tokenIdx == string::npos || tokenIdx >= lineSize) {
+                throw ML::Exception("malformed header");
```
Comment (FinchPowers) [inline http/http_client_impl_v1.cc]: LOL. noexcept mais ça throw?!?

File: http/http_client_impl_v1.cc [hunk 9]
```diff
@@ -0,0 +1,669 @@
...
+    return size;
+}
+
+size_t
+HttpClientImplV1::
+HttpConnection::
+onCurlWrite(const char * data, size_t size)
+    noexcept
+{
+    request_->callbacks()->onData(*request_, data, size);
```
Comment (FinchPowers) [inline http/http_client_impl_v1.cc]: noexcept aussi?

File: http/http_client_impl_v1.h [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+/****************************************************************************/
+
+struct HttpClientImplV1 : public HttpClientImpl {
+    HttpClientImplV1(const std::string & baseUrl,
+                     int numParallel, int queueSize);
+    ~HttpClientImplV1();
+
+    /* AsyncEventSource */
+    virtual int selectFd() const;
+    virtual bool processOne();
```
Comment (FinchPowers) [inline http/http_client_impl_v1.h]: override specifier si je ne me trompe pas?
http://en.cppreference.com/w/cpp/language/override

**Comment by FinchPowers** [comment]
L'entête des fichiers devrait avoir le format suivant.
```
<filename>
<author>, <date>
This file is part of MLDB. Copyright <year> Datacratic. All rights reserved.
```

**Comment by FinchPowers** [comment]
Donc, 
* quelques noexcept à vérifier
* des "override" à ajouter là où applicable
* les entêtes à uniformiser

Beau travail sur les tests. Je suppose que la couverture est bonne?

**Comment by wsourdeau** [comment]
@FinchPowers pour les tests oui, la couverture est pratiquement complète. Chaque bug découvert a apporté son test depuis la version initiale, donc c'est relativement solide.

**Comment by FinchPowers** [comment]
+1

**Pull request merged**